PubSub integration crashing while reading headers

Hi. I just encountered problem with pubsub integration while it's reading headers of message.
I'm using 

```
    def springGcpVersion = "1.0.0.M2"
    compile("org.springframework.cloud:spring-cloud-gcp-pubsub:$springGcpVersion")
    compile("org.springframework.cloud:spring-cloud-gcp-starter-pubsub:$springGcpVersion")
    compile 'org.springframework.integration:spring-integration-core:5.0.1.RELEASE'
```

That's message payload from debug
```
GenericMessage [payload={"challengeId":"XXXXXXXX","RANDOM1":"RANOM","RANDOM2":"SOMETHING","trackingId":"5aXq7iERovyuW4wCdPzxDm","messageType":"SELF"}, headers={spanId=2d9ca7abceb4eda6, spanTraceId=5a8452550ce2c1d353acf05d17aa1147, nativeHeaders={spanTraceId=[5a8452550ce2c1d353acf05d17aa1147], spanId=[2d9ca7abceb4eda6], spanSampled=[0]}, messageSent=true, id=fe942721-1efb-b7ec-854b-cf340c22e5b2, spanSampled=0, gcp_pub_sub_acknowledgement=com.google.cloud.pubsub.v1.MessageDispatcher$3@7871f13d, contentType=text/plain;charset=UTF-8, timestamp=1518622869581}]
```
And that's error value and stacktrace from debug view

```
ex = {ClassCastException@11779} "java.lang.ClassCastException: java.lang.String cannot be cast to java.util.Map"
 detailMessage = "java.lang.String cannot be cast to java.util.Map"
 cause = {ClassCastException@11779} "java.lang.ClassCastException: java.lang.String cannot be cast to java.util.Map"
 stackTrace = {StackTraceElement[37]@11794} 
 suppressedExceptions = {Collections$UnmodifiableRandomAccessList@11387}  size = 0
```

```

 
org.springframework.messaging.handler.annotation.support.HeaderMethodArgumentResolver.getNativeHeaders(HeaderMethodArgumentResolver.java:98)
org.springframework.messaging.handler.annotation.support.HeaderMethodArgumentResolver.getNativeHeaderValue(HeaderMethodArgumentResolver.java:85)
org.springframework.messaging.handler.annotation.support.HeaderMethodArgumentResolver.resolveArgumentInternal(HeaderMethodArgumentResolver.java:69)
org.springframework.messaging.handler.annotation.support.AbstractNamedValueMethodArgumentResolver.resolveArgument(AbstractNamedValueMethodArgumentResolver.java:100)
org.springframework.messaging.handler.invocation.HandlerMethodArgumentResolverComposite.resolveArgument(HandlerMethodArgumentResolverComposite.java:116)
org.springframework.messaging.handler.invocation.InvocableHandlerMethod.getMethodArgumentValues(InvocableHandlerMethod.java:137)
org.springframework.messaging.handler.invocation.InvocableHandlerMethod.invoke(InvocableHandlerMethod.java:109)
org.springframework.integration.util.MessagingMethodInvokerHelper$HandlerMethod.invoke(MessagingMethodInvokerHelper.java:1040)
org.springframework.integration.util.MessagingMethodInvokerHelper.invokeHandlerMethod(MessagingMethodInvokerHelper.java:608)
org.springframework.integration.util.MessagingMethodInvokerHelper.processInternal(MessagingMethodInvokerHelper.java:504)
org.springframework.integration.util.MessagingMethodInvokerHelper.process(MessagingMethodInvokerHelper.java:313)
org.springframework.integration.handler.MethodInvokingMessageProcessor.processMessage(MethodInvokingMessageProcessor.java:104)
org.springframework.integration.handler.ServiceActivatingHandler.handleRequestMessage(ServiceActivatingHandler.java:93)
org.springframework.integration.handler.AbstractReplyProducingMessageHandler.handleMessageInternal(AbstractReplyProducingMessageHandler.java:109)
org.springframework.integration.handler.AbstractMessageHandler.handleMessage(AbstractMessageHandler.java:141)
org.springframework.integration.dispatcher.AbstractDispatcher.tryOptimizedDispatch(AbstractDispatcher.java:116)
org.springframework.integration.dispatcher.UnicastingDispatcher.doDispatch(UnicastingDispatcher.java:132)
org.springframework.integration.dispatcher.UnicastingDispatcher.dispatch(UnicastingDispatcher.java:105)
org.springframework.integration.channel.AbstractSubscribableChannel.doSend(AbstractSubscribableChannel.java:73)
org.springframework.integration.channel.AbstractMessageChannel.send(AbstractMessageChannel.java:438)
org.springframework.integration.channel.AbstractMessageChannel.send(AbstractMessageChannel.java:388)
org.springframework.messaging.core.GenericMessagingTemplate.doSend(GenericMessagingTemplate.java:181)
org.springframework.messaging.core.GenericMessagingTemplate.doSend(GenericMessagingTemplate.java:160)
org.springframework.messaging.core.GenericMessagingTemplate.doSend(GenericMessagingTemplate.java:47)
org.springframework.messaging.core.AbstractMessageSendingTemplate.send(AbstractMessageSendingTemplate.java:108)
org.springframework.integration.endpoint.MessageProducerSupport.sendMessage(MessageProducerSupport.java:197)
org.springframework.cloud.gcp.pubsub.integration.inbound.PubSubInboundChannelAdapter.receiveMessage(PubSubInboundChannelAdapter.java:81)
com.google.cloud.pubsub.v1.MessageDispatcher$4.run(MessageDispatcher.java:405)
java.util.concurrent.Executors$RunnableAdapter.call$$$capture(Executors.java:511)
java.util.concurrent.Executors$RunnableAdapter.call(Executors.java)
java.util.concurrent.FutureTask.run$$$capture(FutureTask.java:266)
java.util.concurrent.FutureTask.run(FutureTask.java)
java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
java.lang.Thread.run(Thread.java:748)
```

I had to copy it via Intellij debug view because at log there was only message without stacktrace. 
Ok I found normal stacktrace

```
org.springframework.messaging.MessageHandlingException: nested exception is java.lang.ClassCastException
	at org.springframework.integration.handler.MethodInvokingMessageProcessor.processMessage(MethodInvokingMessageProcessor.java:107)
	at org.springframework.integration.handler.ServiceActivatingHandler.handleRequestMessage(ServiceActivatingHandler.java:93)
	at org.springframework.integration.handler.AbstractReplyProducingMessageHandler.handleMessageInternal(AbstractReplyProducingMessageHandler.java:109)
	at org.springframework.integration.handler.AbstractMessageHandler.handleMessage(AbstractMessageHandler.java:141)
	at org.springframework.integration.dispatcher.AbstractDispatcher.tryOptimizedDispatch(AbstractDispatcher.java:116)
	at org.springframework.integration.dispatcher.UnicastingDispatcher.doDispatch(UnicastingDispatcher.java:132)
	at org.springframework.integration.dispatcher.UnicastingDispatcher.dispatch(UnicastingDispatcher.java:105)
	at org.springframework.integration.channel.AbstractSubscribableChannel.doSend(AbstractSubscribableChannel.java:73)
	at org.springframework.integration.channel.AbstractMessageChannel.send(AbstractMessageChannel.java:438)
	at org.springframework.integration.channel.AbstractMessageChannel.send(AbstractMessageChannel.java:388)
	at org.springframework.messaging.core.GenericMessagingTemplate.doSend(GenericMessagingTemplate.java:181)
	at org.springframework.messaging.core.GenericMessagingTemplate.doSend(GenericMessagingTemplate.java:160)
	at org.springframework.messaging.core.GenericMessagingTemplate.doSend(GenericMessagingTemplate.java:47)
	at org.springframework.messaging.core.AbstractMessageSendingTemplate.send(AbstractMessageSendingTemplate.java:108)
	at org.springframework.integration.endpoint.MessageProducerSupport.sendMessage(MessageProducerSupport.java:197)
	at org.springframework.cloud.gcp.pubsub.integration.inbound.PubSubInboundChannelAdapter.receiveMessage(PubSubInboundChannelAdapter.java:81)
	at com.google.cloud.pubsub.v1.MessageDispatcher$4.run(MessageDispatcher.java:405)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.ClassCastException: null
```

--- a/spring-cloud-gcp-pubsub/src/main/java/org/springframework/cloud/gcp/pubsub/integration/PubSubHeaderMapper.java
+++ b/spring-cloud-gcp-pubsub/src/main/java/org/springframework/cloud/gcp/pubsub/integration/PubSubHeaderMapper.java
@@ -24,15 +24,16 @@ import org.springframework.cloud.gcp.pubsub.support.GcpPubSubHeaders;
 import org.springframework.integration.mapping.HeaderMapper;
 import org.springframework.integration.util.PatternMatchUtils;
 import org.springframework.messaging.MessageHeaders;
+import org.springframework.messaging.support.NativeMessageHeaderAccessor;
 import org.springframework.util.Assert;
 
 /**
  * Maps headers from {@link com.google.pubsub.v1.PubsubMessage}s to
  * {@link org.springframework.messaging.Message}s and vice-versa.
  *
- * <p>By default, filters out headers called "id", "timestamp" or "gcp_pubsub_acknowledgement" on
- * the {@link org.springframework.messaging.Message} to {@link com.google.pubsub.v1.PubsubMessage}
- * header conversion.
+ * <p>By default, filters out headers called "id", "timestamp", "gcp_pubsub_acknowledgement" or
+ * "nativeHeaders" on the {@link org.springframework.messaging.Message} to
+ * {@link com.google.pubsub.v1.PubsubMessage} header conversion.
  *
  * @author João André Martins
  */
@@ -46,6 +47,7 @@ public class PubSubHeaderMapper implements HeaderMapper<Map<String, String>> {
 			"!" + MessageHeaders.ID,
 			"!" + MessageHeaders.TIMESTAMP,
 			"!" + GcpPubSubHeaders.ACKNOWLEDGEMENT,
+			"!" + NativeMessageHeaderAccessor.NATIVE_HEADERS,
 			"*"};
 
 	/**

--- a/spring-cloud-gcp-pubsub/src/test/java/org/springframework/cloud/gcp/pubsub/integration/PubSubHeaderMapperTests.java
+++ b/spring-cloud-gcp-pubsub/src/test/java/org/springframework/cloud/gcp/pubsub/integration/PubSubHeaderMapperTests.java
@@ -36,7 +36,7 @@ public class PubSubHeaderMapperTests {
 		PubSubHeaderMapper mapper = new PubSubHeaderMapper();
 		Map<String, Object> originalHeaders = new HashMap<>();
 		originalHeaders.put("my header", "pantagruel's nativity");
-		 originalHeaders.put(NativeMessageHeaderAccessor.NATIVE_HEADERS, "deerhunter");
+		originalHeaders.put(NativeMessageHeaderAccessor.NATIVE_HEADERS, "deerhunter");
 		MessageHeaders internalHeaders = new MessageHeaders(originalHeaders);
 
 		Map<String, String> filteredHeaders = new HashMap<>();

--- a/spring-cloud-gcp-pubsub/src/test/java/org/springframework/cloud/gcp/pubsub/integration/PubSubHeaderMapperTests.java
+++ b/spring-cloud-gcp-pubsub/src/test/java/org/springframework/cloud/gcp/pubsub/integration/PubSubHeaderMapperTests.java
@@ -22,6 +22,7 @@ import java.util.Map;
 import org.junit.Test;
 
 import org.springframework.messaging.MessageHeaders;
+import org.springframework.messaging.support.NativeMessageHeaderAccessor;
 
 import static org.assertj.core.api.Assertions.assertThat;
 
@@ -35,6 +36,7 @@ public class PubSubHeaderMapperTests {
 		PubSubHeaderMapper mapper = new PubSubHeaderMapper();
 		Map<String, Object> originalHeaders = new HashMap<>();
 		originalHeaders.put("my header", "pantagruel's nativity");
+		 originalHeaders.put(NativeMessageHeaderAccessor.NATIVE_HEADERS, "deerhunter");
 		MessageHeaders internalHeaders = new MessageHeaders(originalHeaders);
 
 		Map<String, String> filteredHeaders = new HashMap<>();

